### ISSUE TITLE
How to Install

### ISSUE THREAD

**Issue by jeandelacruz** [yaxinr/amqp-mssql-listener#1]
Hi already connect to Queue, but not understand this part : 

![image](https://user-images.githubusercontent.com/7109723/82281829-f5026200-9957-11ea-8b27-322542265b3a.png)

connection : 

![image](https://user-images.githubusercontent.com/7109723/82281908-28dd8780-9958-11ea-9765-182013bda2c7.png)

But no received message.

**Comment by yaxinr** [comment]
Publish message to queue "mssql-subscribe" **FOR SUBSCRIBE TO TABLE CHANGES**

{
  reply_to: "REPLYTO-QUEUE"
  body: 
```
{
  "ConnectionString": "sqlserver://username:password@host:port/instance?database=<Database>",
  "TableName"       : "<Table>",
  "Identity": "a_eq_1",
  "Select": "a",
  "IfUpdate": "update(a) or update(b)",
  "Where": "a='1'",
  "DetailsIncluded": true
}
```
}

**OR**

 create file ./mssql-listeners.json
```
{
    "test": {
        "ConnectionString": "sqlserver://username:password@host:port/instance?database=<Database>",
        "SchemaName": "dbo",
        "TableName"       : "<Table>",
        "Identity": "field1_eq_A",
        "Select": "field1",
        "IfUpdate": "update(field1) or update(field2)",
        "Where": "field1='A'",
        "DetailsIncluded": true
    }
}
```
and bind to exchange "URL.Host-URL.Port-URL.Path-database" with routingKey "TableName-Identity" (URL=ConnectionString)

**Comment by jeandelacruz** [comment]
Hi, Gilfanov

Already configure Queue : 

![image](https://user-images.githubusercontent.com/7109723/82389714-7adae780-9a02-11ea-8c38-ec7da1e58b40.png)

But not listen change to table, my connection is remote.

![image](https://user-images.githubusercontent.com/7109723/82389789-a3fb7800-9a02-11ea-8fe8-b1c5c9bea930.png)

**Comment by yaxinr** [comment]
first read this: https://github.com/dyatchenko/ServiceBrokerListener
"service broker" must be enabled
`ALTER DATABASE MUOnline SET ENABLE_BROKER`

then
"mssql-subscribe" queue not for change messages, it is for subscribe to table

"Change" messages sends to exchange naming like this for your example "localhost-1433-instance-MuOnline".
You can see this exchange in exchanges list.

then you can bind YOUR queue to this exchange by routing key for your example "MEMB_STAT-memb___id" (OR "*" for all tables)

**Comment by jeandelacruz** [comment]
Thanks :), work to Exchange and listen to change in table :)

![image](https://user-images.githubusercontent.com/7109723/82507099-fb631c00-9ac6-11ea-9079-c27d87bb9816.png)

One more question, and how to listen to exchange in nodejs ? this : https://www.rabbitmq.com/tutorials/tutorial-three-javascript.html ?

Greetings (Y)

**Comment by yaxinr** [comment]
your binding is wrong (unbind it).
MUST BE:
"YUOR OWN queue" - "MEMB_STAT-memb___id"

for this in js example:

```js
  var exchange = 'localhost-1433-/instance-MuOnline';

    channel.assertQueue('', {
      exclusive: true
      }, function(error2, q) {
        if (error2) {
          throw error2;
        }
      console.log(' [*] Waiting for logs. To exit press CTRL+C');

      channel.bindQueue(q.queue, exchange, 'MEMB_STAT-memb___id');

      channel.consume(q.queue, function(msg) {
        console.log(" [x] %s: '%s'", msg.fields.routingKey, msg.content.toString());
      }, {
        noAck: true
      });
    });
```

**Comment by jeandelacruz** [comment]
> your binding is wrong (unbind it).
> MUST BE:
> "YUOR OWN queue" - "MEMB_STAT-memb___id"
> 
> for this in js example:
> 
> ```js
>   var exchange = 'localhost-1433-/instance-MuOnline';
> 
>     channel.assertQueue('', {
>       exclusive: true
>       }, function(error2, q) {
>         if (error2) {
>           throw error2;
>         }
>       console.log(' [*] Waiting for logs. To exit press CTRL+C');
> 
>       channel.bindQueue(q.queue, exchange, 'MEMB_STAT-memb___id');
> 
>       channel.consume(q.queue, function(msg) {
>         console.log(" [x] %s: '%s'", msg.fields.routingKey, msg.content.toString());
>       }, {
>         noAck: true
>       });
>     });
> ```

thank you for the pacience, so binding :

![image](https://user-images.githubusercontent.com/7109723/82524312-3891d300-9af4-11ea-890b-a398068a183a.png)

this is created by running the node example and : 

![image](https://user-images.githubusercontent.com/7109723/82524334-4c3d3980-9af4-11ea-85b9-3af52d3e4c14.png)

So where would the binding go?

Thanks (Y)

**Comment by yaxinr** [comment]
> So where would the binding go?

i dont undertand.
in this code queue is temporary exclusive for this app instance. when app is closed queue is deleted.

for persisted you can declare queue with name
channel.assertQueue('my queue for table changes', ...

**Comment by jeandelacruz** [comment]
> > So where would the binding go?
> 
> i dont undertand.
> in this code queue is temporary exclusive for this app instance. when app is closed queue is deleted.
> 
> for persisted you can declare queue with name
> channel.assertQueue('my queue for table changes', ...

Of course in that part I can understand and there is no problem, 

![image](https://user-images.githubusercontent.com/7109723/82525516-2cf3db80-9af7-11ea-8b32-059e53523e9d.png)

This is what I can't understand very well.

**Comment by yaxinr** [comment]
this binding is wrong
![image](https://user-images.githubusercontent.com/1458827/82526186-c7323e80-9b4c-11ea-9a49-8184861e5411.png)

**Comment by jeandelacruz** [comment]
> this binding is wrong
> ![image](https://user-images.githubusercontent.com/1458827/82526186-c7323e80-9b4c-11ea-9a49-8184861e5411.png)

yes, this part I understand, what I did not understand was: "YUOR OWN queue" - "MEMB_STAT-memb___id".

thanks for the patience

**Comment by yaxinr** [comment]
you must declare **your** queue, then bind this queue to exchange...

**Comment by jeandelacruz** [comment]
Hi, according to what you tell me, should I create a new queue?
then be able to add on the exchange?

Thanks :)

**Comment by yaxinr** [comment]
> Hi, according to what you tell me, should I create a new queue?
> then be able to add on the exchange?
> 
> Thanks :)

if you should consume all table changes, then you should create a new persisted queue.

**Comment by longyn** [comment]
Hi @jeandelacruz,

Met same problem, could you share your configuration?
